### ISSUE TITLE
gatsby-plugin-netlify-cms webpack config issues

### ISSUE THREAD

**Issue by S-unya** [netlify/netlify-cms#1318]
**- Do you want to request a *feature* or report a *bug*?**

*bug*

**- What is the current behavior?**

When running `gatsby build` webpack fails with: 
`There were errors with your webpack config:
[1]
module.loaders.[indexOfNetlifyCMSPlugin].loader
string.base , "loader" must be a string`

**NOTE:** This doesn't seem to be a problem for `gatsby develop`.

**- If the current behavior is a bug, please provide the steps to reproduce.**

As far as I can tell all you need is:
1. `"gatsby-plugin-netlify-cms": "^1.0.12",` in package.json
2. run `npm i` or `yarn install`
3. run `npm run build` or `gatsby build`

**- What is the expected behavior?**

I expect the webpack build phases to all complete without errors

**- Please mention your versions where applicable.**

Netlify CMS version: any > 1.1.0
Browser version: n/a

Node.JS version: v9.11.1
Operating System: Mac OSX 0.13.5 Beta

**- Please link or paste your `config.yml` below if applicable.**
![my config.yml](https://github.com/aerian-studios/gatsby-netlify-poc/blob/master/static/admin/config.yml)

**- Further notes.**
I had play with modifying the webpack config via `modifyWebpackConfig`, because I could see that the `loader` parameter was in the format, `Array<{loader: string; options: {};}>` rather than the expected `string`. I only wanted to verify that providing a string for this parameter solves the first issue:
```javascript
exports.modifyWebpackConfig = ({ config, stage }) => {
  if (stage === "build-javascript") {
    config.loader("cms-css", current => {
      const opts = current.loader[0].options;
      const str = `{"id":${opts.id}, "remove":${opts.remove}, "omit":${
        opts.omit
      }}`;

      return {
        test: current.test,
        include: current.include,
        loader: `${current.loader[0].loader}?${str}!css`
      };
    });
  }
};
```
The above allows webpack to progress further but causes other issues, probably because I'm only outputting the one loader. I will probably try to convert it to use the `loaders` param next, to see if that solves all problems.

I can try provide more info/help if anyone else is getting this is is able to reproduce.

**Comment by S-unya** [comment]
For further information, which may be obvious to someone more aufait with webpack config; I mapped the `loader` array to the `loaders` parameter and got a new error:

```
 Building production JavaScript bundlests-loader: Using typescript@2.8.1 and /*/workspace/gatsby-netlify-poc/tsconfig.json
error UNHANDLED EXCEPTION


  TypeError: chunk.isInitial is not a function

  - helpers.js:15 isInitialOrHasNoParents
    [gatsby-netlify-poc]/[extract-text-webpack-plugin]/dist/lib/helpers.js:15:16

  - index.js:187
    [gatsby-netlify-poc]/[extract-text-webpack-plugin]/dist/index.js:187:94

  - async.js:3096
    [gatsby-netlify-poc]/[async]/dist/async.js:3096:16

  - async.js:1055 eachOfArrayLike
    [gatsby-netlify-poc]/[async]/dist/async.js:1055:9

  - async.js:1103 eachOf
    [gatsby-netlify-poc]/[async]/dist/async.js:1103:5

  - async.js:3158 Object.eachLimit
    [gatsby-netlify-poc]/[async]/dist/async.js:3158:5

  - index.js:184 Compilation.<anonymous>
    [gatsby-netlify-poc]/[extract-text-webpack-plugin]/dist/index.js:184:27

  - Tapable.js:69 Compilation.next
    [gatsby-netlify-poc]/[tapable]/lib/Tapable.js:69:14

  - index.js:285 ExtractTextPlugin.<anonymous>
    [gatsby-netlify-poc]/[gatsby-1-config-extract-plugin]/[extract-text-webpack-plugin]/index.js:285:5

  - async.js:52
    [gatsby-netlify-poc]/[gatsby-1-config-extract-plugin]/[async]/lib/async.js:52:16

  - async.js:236 Object.async.forEachOf.async.eachOf
    [gatsby-netlify-poc]/[gatsby-1-config-extract-plugin]/[async]/lib/async.js:236:30

  - async.js:209 Object.async.forEach.async.each
    [gatsby-netlify-poc]/[gatsby-1-config-extract-plugin]/[async]/lib/async.js:209:22

  - index.js:237 ExtractTextPlugin.<anonymous>
    [gatsby-netlify-poc]/[gatsby-1-config-extract-plugin]/[extract-text-webpack-plugin]/index.js:237:10

  - Tapable.js:71 Compilation.applyPluginsAsync
    [gatsby-netlify-poc]/[tapable]/lib/Tapable.js:71:13

  - Compilation.js:525 Compilation.seal
    [gatsby-netlify-poc]/[webpack]/lib/Compilation.js:525:7

  - Compiler.js:397 Compiler.<anonymous>
    [gatsby-netlify-poc]/[webpack]/lib/Compiler.js:397:15
```

and this is what I did to map:

``` javascript
exports.modifyWebpackConfig = ({ config, stage }) => {
  if (stage === "build-javascript") {
    config.loader("cms-css", current => {
      const constructedLoaders = current.loader.map(loader => ({
        loader: loader.loader,
        query: loader.options ? loader.options : {}
      }));
      const newCurrent = {
        test: current.test,
        include: current.include,
        loaders: constructedLoaders
      };

      return newCurrent;
    });
  }
};
```

**Comment by erquhart** [comment]
Hmm wonder if this is due to your cms script being written in Typescript - can you test with a js variant of your cms.js?

**Comment by S-unya** [comment]
Hi @erquhart, thanks for the speedy reply.

I altered `cms.ts` to `cms.js`, I also adjusted the files referenced within it to `.jsx` (as an separate test) and modified `gatsby-config.js` to point at the js file and had the same behaviours as above. It is on this branch, https://github.com/aerian-studios/gatsby-netlify-poc/tree/gatsby-cms-debugging.

I think that there may be 2 issues, but I think that one is a loader config conflict. My personal bet is the *extract-text-webpack-plugin*, but it would be trivial to switch all the files to js if you think the issue is with *ts-loader*.

One important thing to note is that this worked previously and it stopped working after running `npm i`. I verified this on another computer that had the project installed. It ran through the build process correctly. I then ran `npm i` and it stopped being able to run through, showing the above errors.

**Comment by erquhart** [comment]
Ah, didn't realize you were modifying Gatsby's webpack, that can definitely introduce conflicts.

Need to clarify two things:

1. Have you tested without the netlify cms plugin to ensure that it's the problem?
2. If so, can you determine the last version of the plugin that doesn't break your project?

**Comment by S-unya** [comment]
I added the mods to try to overcome the issues I was having.

I can run the build happily if I comment out the netlify-cms plugin.

I'll get back to you on the version question (2 above)

**Comment by S-unya** [comment]
@erquhart, version `@1.0.02` is the highest version that will successfully build... That was a fascinating hour decrementing versions and testing build 😄

**Comment by S-unya** [comment]
Not sure if it is useful, but I had a quick step through of the webpack loaders and the most obvious difference is no `cms-css` loader in this lower version, which was the loader that was creating those errors for me in later versions.

**Comment by erquhart** [comment]
Thanks for taking the time to step through that! You should checkout [`git bisect`](https://robots.thoughtbot.com/git-bisect) next time you need to test something like this, it's a game changer for this sort of debugging.

I cloned your repo and can't get it to build even with your latest commit reverting the cms plugin version. The error is `Error: Chunk.entry was removed. Use hasRuntime()`, and it's coming right after the typescript loader logs it's version to console. What's curious is that the typescript loader is being used during the CSS build phase - looking at the plugin, it runs on every phase instead of just the js phases.

I'm betting the cause is a conflict between the Netlify CMS plugin and the TypeScript plugin, and I'm wondering if simply limiting the ts plugin to js phases will do the trick. Regardless, this is beyond Netlify CMS core - can you open an issue in the Gatsby repo?

**Comment by S-unya** [comment]
@erquhart thanks for the pointer regarding `git bisect`. I will investigate your ideas some more, and if I get no joy, I will open an issue on Gatsby.

Once again, thanks for your time and hard work.

**Comment by S-unya** [comment]
Further to this thread. I decided to try investigating this a bit further before going on. After some investigation following the css/ts compilation thread I discovered that there was definitely an issue there (I experimented with removing postcss-sass and typescript, etc), but I was unable to pinpoint it exactly...

That notwithstanding, I experimented with installing everything from scratch, in order of importance for me and everything simply worked! So, unsatisfying in that I never managed to pinpoint what went wrong, but great in that I have everything working. 

Considering that this is probably the consequence of order of install or somesuch, perhaps this should be closed?
